### ISSUE TITLE
storage: log disk stalls to system.eventlog

### ISSUE THREAD

**Issue by mwang1026** [cockroachdb/cockroach#125977]
Perusing the "events" window in the DB console is a common initial flow when diagnosing issues to make sure nothing obvious correlates with timeline of an issue. Disk slowness and stalls are not currently logged in the event view (which is sourced from `system.eventlog` table). Currently disk slowness and stalls are only logged in pebble logs.

Ask here is that the we log an event into `system.eventlog`. Log when the slowness is first detected, and then when it clears log the event:

- Timestamp
- EventType: disk_slowness_detected; disk_slowness_cleared
- NodeID: node
- StoreID: store_id

Note:  Writing to the event log in these cases should be best effort, as if there's a disk stall, and the leaseholder for the event log is on the same node, we think it might deadlock. 

Jira issue: CRDB-39694

**Comment by itsbilal** [comment]
Currently the EventListener API between Pebble and Cockroach for disk slowness events looks like this:

https://github.com/cockroachdb/cockroach/blob/9af051e4151254a8ef10ce50a0990f7c26605226/pkg/storage/pebble.go#L1593

Usually when there's a slow disk, we see a splutter of calls to this method over the slowness period of time. We could either change this API on Pebble's side to explicitly emit start/end events (with the ends not being guaranteed in case of a crash, obviously), or we could coalesce these events into a start/end event on the Cockroach side.

In the interest of keeping the Pebble-side disk health checks as simple as possible so they catch all instances of slownesses, I'd be in favour of doing this coalescing/deduplication on the Cockroach side, where a store keeps track of whether it's in a slowness event already (if not, it creates one) when it sees a Pebble DiskSlow, then 20 seconds after the last disk slowness event from Pebble, we mark an event on that store as having ended. The 20 seconds could be configurable, or it could reuse the same MaxSyncDuration cluster setting as disk stalls.

**Comment by RaduBerinde** [comment]
@mwang1026 We're trying to come up with more general ways to do this kind of thing and I had a few questions. If a node is hitting slowness events for long periods of time, it is possible that the start event is old and not noticed anymore. Would it be useful to periodically post a "still happening" event? We're also thinking about perhaps assigning a period of time to each event, and showing all intersecting events for the current metrics timeframe. In that world, we would periodically post "disk slowness detected" events which contain a specific timeframe (and no "cleared" event).

Basically I want to see if I'm missing an important customer perspective or requirement, or if it's ok to make whatever decisions here we feel make for the best usability.

**Comment by mwang1026** [comment]
> If a node is hitting slowness events for long periods of time, it is possible that the start event is old and not noticed anymore.

How old? My understanding is we don't fire off that many alerts to system.event_log that 20s will have that alert buried deep, short of some major schema changes or something going on at the same time, but I can check a few cloud clusters to validate that.

>  Would it be useful to periodically post a "still happening" event?
That could be an alternative approach. Let me ask a few folks -- the core issue here is correlating disk stall/slowness with other symptoms, usually sql thruput/latency blips -- and it might actually be more useful to have the periodic ping vs. having to find start and stop, and even those starts and stops being somewhat inaccurate (in that it's subject to how often we check for a stall). With a periodic check you could just assume the first and last that you see in a time frame, +/- some amount, is how long the disk was stalled. The only tricky part with this would be "flopping" disks but I'm not sure start/stop variant would be better in that instance either. Thoughts?
